exercises/bank-account/BankAccountSpec.groovy

```diff
@@ -0,0 +1,202 @@
+    def cannotWithdrawMoneyFromEmptyAccount() {
+        setup:
+            BankAccount bankAccount=new BankAccount()
+
+        when:         
+            bankAccount.withdraw(5);
+        then:
+            thrown(Exception)    
+    }
+
+    def cannotWithdrawMoreMoneyThanYouHave() {
+        setup:
+            BankAccount bankAccount=new BankAccount()
+        when:
+            bankAccount.open();
+            bankAccount.deposit(6);
+            bankAccount.withdraw(7);
+        then:
+            thrown(Exception)           
+    }
+
+    def cannotDepositNegativeAmount() {
+        setup:
+            BankAccount bankAccount=new BankAccount()
+        when:
+            bankAccount.open();
+            bankAccount.deposit(-1);
+
+        then:
+            thrown(Exception)                   
+    }
+
+    def cannotWithdrawNegativeAmount() {
+        setup:
+            BankAccount bankAccount=new BankAccount()
+        when:
+            bankAccount.open();
+            bankAccount.deposit(105);
+            bankAccount.withdraw(-5);
+
+        then:
+            thrown(Exception)                   
+    }
+
+    def cannotGetBalanceOfClosedAccount() {
+        setup:
+            BankAccount bankAccount=new BankAccount()
+        when:
+            bankAccount.open();
+            bankAccount.deposit(10);
+            bankAccount.close();
+            bankAccount.getBalance();
+
+        then:
+            thrown(Exception)                   
+    }
+
+    def cannotDepositMoneyIntoClosedAccount() {
+        setup:
+            BankAccount bankAccount=new BankAccount()
+        when:
+            bankAccount.open();
+            bankAccount.close();
+            bankAccount.deposit(5);
+
+        then:
+            thrown(Exception)                   
+    }
+
+    def cannotWithdrawMoneyFromClosedAccount() {
+        setup:
+            BankAccount bankAccount=new BankAccount()
+        when:
+            bankAccount.open();
+            bankAccount.deposit(20);
+            bankAccount.close();
+            bankAccount.withdraw(5);
+
+        then:
+            thrown(Exception)                   
+    }
+
+    def bankAccountIsClosedBeforeItIsOpened() {
+        setup:
+            BankAccount bankAccount=new BankAccount()
+        when:
+            bankAccount.getBalance();
+
+        then:
+            thrown(Exception)                   
+    }
+
+    def canAdjustBalanceConcurrently() {
+        setup:
+            BankAccount bankAccount=new BankAccount();
+        when:
+            bankAccount.open();
+            bankAccount.deposit(1000);
+            for (int i = 0; i < 10; i++) {
+                adjustBalanceConcurrently(bankAccount);
+            }
+        
+        then:
+            bankAccount.getBalance() == 1000
+    }
+
+    void adjustBalanceConcurrently(BankAccount bankAccount) {
+        Random random = new Random();
+
+        Thread[] threads = new Thread[1000];
+        for (int i = 0; i < 1000; i++) {
```

Comment:
It's possible to make the code more Groovish by using constructions like closures etc:
````groovy
(1..1000).each { println it }
````